### ISSUE TITLE
Reproducing Fig. 3 

### ISSUE THREAD

**Issue by JeanRadig** [EperLuo/scDiffusion#5]
Hi! 

I am trying to reproduce Fig. 3, in particular the umap of muris with the different cell types. 

<img width="646" alt="Screenshot 2024-06-05 at 15 16 28" src="https://github.com/EperLuo/scDiffusion/assets/44983330/35adcb8d-8411-49cd-bd5f-c47ad7aad147">

I have trained the models on the muris dataset. 
I try to get the conditional samples. 
I do not understand what parameters I should pass in the function main of classifier_sample.py to get the 12 samples, as needed in Jupyter notebook "script_diffusion_map", lines:

---- ---- ---- ---- ---- ---- ---- ---- ---- 
index2 = [0,1,2,3,4,5,6,7,8,9,10,11]
for i in index2:
    npzfile=np.load(f'/data1/lep/Workspace/guided-diffusion/output/muris_condi/muris_{i}_scimilarity.npz',allow_pickle=True)
---- ---- ---- ---- ---- ---- ---- ---- ---- 

Could you please let me know how you have done it?

**Comment by JeanRadig** [comment]
1. **Update train.sh**. You can modify train.sh to accommodate for the right arguments.

- Modified train.sh in function of the arguments that can actually be passed, as defined for each function. Previous commands could not run due to unexpected arguments such as e.g. --num_genes for the diffusion and classifier. Also modified the arguments name, when they did not correspond to those as defined in the functions.

```
cd VAE
echo "Training Autoencoder, this might take a long time"
CUDA_VISIBLE_DEVICES=0 python VAE_train.py --data_dir 'path/to/anndata.h5ad' --num_genes 18996 --save_dir 'path/to/saved_VAE_model' --max_steps 200000
echo "Training Autoencoder done"

cd ..
echo "Training diffusion backbone"
CUDA_VISIBLE_DEVICES=0 python cell_train.py --data_dir 'path/to/anndata.h5ad'  --vae_path 'path/to/saved_VAE_model/VAE_checkpoint.pt' \
    --save_dir 'path/to/saved_diffusion_model' --model_name 'my_diffusion' --save_interval 20000
echo "Training diffusion backbone done"

echo "Training classifier"
CUDA_VISIBLE_DEVICES=0 python classifier_train.py --data_dir 'path/to/anndata.h5ad' --model_path "path/to/saved_classifier_model" \
    --iterations 40000 --vae_path 'path/to/saved_VAE_model/VAE_checkpoint.pt'
echo "Training classifier, done"
```
3. **Guide the user**. You can insert some more explanation at the beginning of your notebooks to guide the user. Example on top of script_diffusion_umap.ipynb:

Before running the following code, ensure that you have trained the VAE model, the diffusion model, and the classifier model.

Given an AnnData file you wish to analyze, named `anndata.h5ad`, please follow these steps in your command line to create the necessary input data:

1. **Train VAE**:
   - Prior to training, create the folder `path/to/saved_VAE_model`.
   - In the terminal, navigate to the `VAE` directory:
     ```bash
     cd VAE
     ```
   - Run the following command to train the Autoencoder:
     ```bash
     echo "Training Autoencoder, this might take a long time" 
     CUDA_VISIBLE_DEVICES=0 python VAE_train.py --data_dir 'path/to/anndata.h5ad' --num_genes 18996 --save_dir 'path/to/saved_VAE_model' --max_steps 200000
     echo "Training Autoencoder done"
     ```

3. **Train the Diffusion Model**:
   - Prior to training, create the folder `path/to/saved_diffusion_model`.
   - In the terminal, navigate back to the root directory:
     ```bash
     cd ..
     ```
   - Run the following command to train the diffusion backbone:
     ```bash
     echo "Training diffusion backbone"
     CUDA_VISIBLE_DEVICES=0 python cell_train.py --data_dir 'path/to/anndata.h5ad'  --vae_path 'path/to/saved_VAE_model/VAE_checkpoint.pt' \
         --save_dir 'path/to/saved_diffusion_model' --model_name 'my_diffusion' --save_interval 20000
     echo "Training diffusion backbone done"
     ```

4. **Train the Classifier**:
   - Prior to training, create the folder `path/to/saved_classifier_model`.
   - Run the following command to train the classifier:
     ```bash
     echo "Training classifier"
     CUDA_VISIBLE_DEVICES=0 python classifier_train.py --data_dir 'path/to/anndata.h5ad' --model_path "path/to/saved_classifier_model" \
         --iterations 40000 --vae_path 'path/to/saved_VAE_model/VAE_checkpoint.pt'
     echo "Training classifier, done"
     ```

Once the models are trained, you need to create the `.npz` files which will serve as input for the following steps. To do this:

1. **Unconditional Sampling (Generate Data from Diffusion Model)**:
   - Prior to sampling, create the folder `path/to/saved_unconditional_sampling`.
   - Run the following command to perform unconditional sampling:
     ```bash
     # Unconditional sampling
     python cell_sample.py --model_path "path/to/saved_diffusion_model/checkpoint.pt" --sample_dir "path/to/saved_unconditional_sampling"
     ```

2. **Conditional Sampling (Generate Data from Classifier Model)**:
   - Prior to sampling, create the folder `path/to/saved_conditional_sampling`.
   - You also need to modify the `main()` function in `classifier_sample.py` to create the samples based on your specified condition. **NEEDS MORE DOCUMENTATION: PROVIDE CLEAR EXAMPLE**
   - Run the following command to perform conditional sampling:
     ```bash
     # Conditional sampling 
     python classifier_sample.py --model_path "path/to/saved_diffusion_model/checkpoint.pt" --classifier_path "path/to/saved_classifier_model/checkpoint.pt" --sample_dir "path/to/saved_conditional_sampling"
     ```

Ensure that you replace `'path/to/anndata.h5ad'`, `'path/to/saved_VAE_model'`, `'path/to/saved_diffusion_model'`, and `'path/to/saved_classifier_model'` with the actual paths in your system. Additionally, make sure to adjust any other parameters according to your specific setup and requirements.

**Comment by JeanRadig** [comment]
[0001-Modified-train.sh-in-function-of-the-arguments-that-.patch](https://github.com/user-attachments/files/15604891/0001-Modified-train.sh-in-function-of-the-arguments-that-.patch)

**Comment by Sbs12** [comment]
Hello, can your unconditional generation effect achieve the effect described in the article?

**Comment by EperLuo** [comment]
Hi @JeanRadig, sorry for the late reply. This work is currently under revision. We will update the code repository to give a more comprehensive guidance for the user. 

As for the conditional generation problem, you need to change the code in the main() to :
`if __name__ == "__main__":
    # for conditional generation
    # main(cell_type=[2])
    for type in range(12):#[0,1,2,3,4,5]:
        main(cell_type=[type])
`
to generate all 12 types of cells. And also change the `sample_dir=f"output/simulated_samples/muris",` in line 310 to `sample_dir=f"output/simulated_samples/muris_{celltype[0]}",` to save them. 

We would update these files these days. Thank you very much for those comments!

**Comment by JeanRadig** [comment]
@Sbs12, I do not manage to reproduce the results as of now. For what I have been trying lately, neither the conditional nor the unconditional data generation worked. In particular, the gradient interpolation yields batches of data that are uncorrelated to the true data, and undistinguishable from one another on a UMAP. But the authors are making updates so we need to stay tuned!

@EperLuo, thank you very much for your answer! I indeed managed to find out what to change. I also wanted to post here the results of my trial to reproduce the gradient interpolation, which I think is a very cool idea. I did not manage to obtain the desired results, could you tell me why it is so?

Using the following data as original: 

### Original data:

![image](https://github.com/EperLuo/scDiffusion/assets/44983330/c0ab487b-8961-4993-b2a2-1b6cde026c2c)

And the following as training data.

### **Training data (original minus D3 - D4.5)**:

![image](https://github.com/EperLuo/scDiffusion/assets/44983330/3c59ed7c-dcf7-48ee-bad2-c2644dc2095e)

We train the models.

## Training steps

### Training scimilarity

```python
cd VAE
echo "Training Autoencoder, this might take a long time"
CUDA_VISIBLE_DEVICES=0 python VAE_train.py --data_dir '/workspace/projects/001_scDiffusion/data/data_in/data/WOT/train_WOT.h5ad' --num_genes 19423 --state_dict "/workspace/projects/001_scDiffusion/scripts/scDiffusion/annotation_model_v1" --save_dir '../checkpoint/AE/WOT_VAE' --max_steps 200000 --max_minutes 60
echo "Training Autoencoder done"
```

First, we need to do the following change on top of **VAE_train.py** to specify we are loading WOT data:

```python
# from guided_diffusion.cell_datasets import load_data
# from guided_diffusion.cell_datasets_sapiens import load_data
from guided_diffusion.cell_datasets_WOT import load_data
# from guided_diffusion.cell_datasets_muris import load_data
```

(note that when we were dealing with muris dataset it was muris).

### Training diffusion

```python
cd ..
echo "Training diffusion backbone"
CUDA_VISIBLE_DEVICES=0 python cell_train.py --data_dir '/workspace/projects/001_scDiffusion/data/data_in/data/WOT/train_WOT.h5ad'  --vae_path '/workspace/projects/001_scDiffusion/scripts/scDiffusion/checkpoint/AE/WOT_VAE/model_seed=0_step=163856.pt' \
    --save_dir '/workspace/projects/001_scDiffusion/scripts/output/WOT_diffusion' --model_name 'my_diffusion' --save_interval 20000
echo "Training diffusion backbone done"
```

Where on top of cell_train.py we need to uncomment/comment as follows:

```python
# In cell_train.py

from guided_diffusion import dist_util, logger
# from guided_diffusion.cell_datasets import load_data
 from guided_diffusion.cell_datasets_WOT import load_data
# from guided_diffusion.cell_datasets_sapiens import load_data
# from guided_diffusion.cell_datasets_muris import load_data
from guided_diffusion.resample import create_named_schedule_sampler
```

### Training classifier

```python
echo "Training classifier"
CUDA_VISIBLE_DEVICES=0 python classifier_train.py --data_dir '/workspace/projects/001_scDiffusion/data/data_in/data/WOT/train_WOT.h5ad' --model_path "/workspace/projects/001_scDiffusion/scripts/output/WOT_classifier" \
    --iterations 100000 --vae_path '/workspace/projects/001_scDiffusion/scripts/scDiffusion/checkpoint/AE/WOT_VAE/model_seed=0_step=163856.pt'
echo "Training classifier, done"
```

Where on top of classifier_train.py we need to change following command:

```python
# In classifier_train.py

from guided_diffusion import dist_util, logger
from guided_diffusion.fp16_util import MixedPrecisionTrainer
# from guided_diffusion.cell_datasets_pbmc import load_data
from guided_diffusion.cell_datasets_WOT import load_data
# from guided_diffusion.cell_datasets_muris import load_data
# from guided_diffusion.cell_datasets_sapiens import load_data
```

And change the number of classes in:

```python
# In classifier_train.py, function create_argparser()

defaults.update(classifier_and_diffusion_defaults())
defaults['num_class']= 15  # <----- Change the number of classes to 15
parser = argparse.ArgumentParser()
add_dict_to_argparser(parser, defaults)
return parser
```

### Creating latent spaces

Note that for scDiffusion, a few more arguments are needed, as can be found in classifier_sample.py:

```python
# In classifier_sample.py 
# if gradient interpolation
ae_dir='output/Autoencoder_checkpoint/WOT/model_seed=0_step=150000.pt', 
num_gene=19423, # WOT 19423
init_time = 600,    # initial noised state if interpolation
init_cell_path = 'data/WOT/filted_data.h5ad',   #input initial noised cell state

sample_dir=f"output/simulated_samples/muris",
start_guide_steps = 500,     # the time to use classifier guidance
filter = False,   # filter the simulated cells that are classified into other condition, might take long time
```

```python
# Conditional sampling 
python classifier_sample.py --model_path "/workspace/projects/001_scDiffusion/scripts/output/WOT_diffusion/my_diffusion/model200000.pt" --classifier_path "/workspace/projects/001_scDiffusion/scripts/output/WOT_classifier/model099999.pt" --sample_dir "/workspace/projects/001_scDiffusion/scripts/output/WOT_simulated" --ae_dir '/workspace/projects/001_scDiffusion/scripts/scDiffusion/checkpoint/AE/WOT_VAE/model_seed=0_step=163856.pt' --init_cell_path '/workspace/projects/001_scDiffusion/data/data_in/data/WOT/filted_data.h5ad' --num_gene 19423 
```

Where we need to change the number of classes to here 15 (17 - 2, e.g. 3.5 and 4):

```python
# In classifier_sample.py, funciton create_argparser()

defaults.update(model_and_diffusion_defaults())
defaults.update(classifier_and_diffusion_defaults())
defaults['num_class']=15 # <--------- Change the number of classes to 15
parser = argparse.ArgumentParser()
add_dict_to_argparser(parser, defaults)
return parser
```

We are going to simulate:

```python
if __name__ == 'main':
	parser = create_argparser()
  args = parser.parse_args()
  save_dir = args.sample_dir
  for i in range(0,11):
      path_to_save = save_dir + f"{i}"
      to_save = main(cell_type=[6,7], inter=True, weight=[10-i,i])
      save_data(to_save, i, path_to_save)
```

We are simulating 11 samples between the cell types 6 (3) and 7 (4.5) of the filtered anndata. 

### Visualising results

Decoding the data necessitates to also change the num_genes to 19423 in:

```python
def load_VAE():
    autoencoder = VAE(
        num_genes=19423,
        device='cuda',
        seed=0,
        loss_ae='mse',
        hidden_dim=128,
        decoder_activation='ReLU',
    )
    autoencoder.load_state_dict(torch.load('/workspace/projects/001_scDiffusion/scripts/scDiffusion/checkpoint/AE/WOT_VAE/model_seed=0_step=163856.pt'))
    return autoencoder
```

I obtain:

![image](https://github.com/EperLuo/scDiffusion/assets/44983330/10e5b829-a52f-4264-b615-4b7ac3da5726)


But they all overlap:

![image](https://github.com/EperLuo/scDiffusion/assets/44983330/9a69b577-707a-4da8-833e-cd2ec585e03d)


And they do not correlate with the training data:

![image](https://github.com/EperLuo/scDiffusion/assets/44983330/e79e524c-990b-4c1b-8d27-2df4c56d7efe)

### What modifications should I do to obtain the desired results?

**Comment by EperLuo** [comment]
Hi, very glad you're so interested in our work! The training process seems correct, but to further determine the training and inference process, could you try generating the known cells (like all cell types from day 0 to day 8) and check if these cells can overlap with the real cells in the UMAP? This could help to confirm the proper usage of the model.

Besides, I do recommend you use the exp_script/script_diffusion_interpolation.ipynb to reproduce the results in the paper. The UMAP visual results might be different under different cell numbers. If the generated cells are way more than the real cells, the UMAP might be looked at as incorrelated. We set the number of generated cells in this script so that it is of the same order of magnitude as the number of real cells, and added real known cells to the decoder to balance out the effect of the layernorm in SCimilarity. I think this might help you to reproduce the results.
